File: src/main/java/com/mhl/mycompanybackend/userdetails/UserDetailsImpl.java
package com.mhl.mycompanybackend.userdetails;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.mhl.mycompanybackend.model.Roles;
import com.mhl.mycompanybackend.model.Users;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.io.Serial;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class UserDetailsImpl implements UserDetails {

    @Serial
    private static final long serialVersionUID = 1L;

    private final Long id;
    private final String username;
    private final String email;
    @JsonIgnore
    private String password;
    private final Collection<? extends GrantedAuthority> authorities;
    private final List<Roles> roles;

    public UserDetailsImpl(Long id, String username, String email, String password,
                           Collection<? extends GrantedAuthority> authorities, List<Roles> roles) {
        this.id = id;
        this.username = username;
        this.email = email;
        this.password = password;
        this.authorities = authorities;
        this.roles = roles;
    }

    public static UserDetailsImpl build(Users users) {
        List<GrantedAuthority> authorities = users.getRoles().stream()
                .map(role -> new SimpleGrantedAuthority(role.getPermissions().getName().name()))
                .collect(Collectors.toList());

        return new UserDetailsImpl(
                users.getId(),
                users.getUsername(),
                users.getEmail(),
                users.getPassword(),
                authorities, users.getRoles());
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public String getUsername() {
        return username;
    }

    public Long getId() {
        return id;
    }

    public String getEmail() {
        return email;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    public List<Roles> getRoles() {
        return roles;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((id == null) ? 0 : id.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        UserDetailsImpl other = (UserDetailsImpl) obj;
        if (id == null) {
            return other.id == null;
        } else return id.equals(other.id);
    }

}

File: src/main/java/com/mhl/mycompanybackend/userdetails/UserDetailsImpl.kt
package com.mhl.mycompanybackend.userdetails

import com.fasterxml.jackson.annotation.JsonIgnore
import com.mhl.mycompanybackend.model.Roles
import com.mhl.mycompanybackend.model.UserData
import com.mhl.mycompanybackend.model.Users
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.io.Serial
import java.util.stream.Collectors

class UserDetailsImpl(
    val id: Long?, private val username: String, val email: String, @field:JsonIgnore private val password: String,
    private val authorities: Collection<GrantedAuthority>, val roles: List<Roles>, val userData: UserData
) : UserDetails {

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return authorities
    }

    override fun getPassword(): String {
        return password
    }

    override fun getUsername(): String {
        return username
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return true
    }

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + (id?.hashCode() ?: 0)
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) return true
        if (obj == null) return false
        if (javaClass != obj.javaClass) return false
        val other = obj as UserDetailsImpl
        return if (id == null) {
            other.id == null
        } else id == other.id
    }

    companion object {
        @Serial
        private val serialVersionUID = 1L
        fun build(users: Users): UserDetailsImpl {
            val authorities = users.roles.stream()
                .map { role: Roles -> SimpleGrantedAuthority(role.permissions!!.name!!.name) }
                .collect(Collectors.toList())
            println(users.userData!!.firstname)
            return UserDetailsImpl(
                users.id,
                users.username!!,
                users.email!!,
                users.password!!,
                authorities, users.roles, users.userData!!
            )

//
        }
    }
}
